bastionzero/autodiscoveryscript/datasource_adbash_test.go

```diff
@@ -0,0 +1,75 @@
+package autodiscoveryscript_test
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"testing"
+
+	"github.com/bastionzero/bastionzero-sdk-go/bastionzero/service/autodiscoveryscripts/targetnameoption"
+	"github.com/bastionzero/bastionzero-sdk-go/bastionzero/service/policies"
+	"github.com/bastionzero/terraform-provider-bastionzero/internal/acctest"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+)
+
+func TestAccADBashDataSource_Basic(t *testing.T) {
+	ctx := context.Background()
+	dataSourceName := "data.bastionzero_ad_bash.test"
+	env := new(policies.Environment)
+
+	acctest.SkipIfNotInAcceptanceTestMode(t)
+	acctest.PreCheck(ctx, t)
+	acctest.FindNEnvironmentsOrSkip(t, env)
+
+	makeTestStep := func(targetNameOption targetnameoption.TargetNameOption) resource.TestStep {
+		return resource.TestStep{
+			Config: testAccADBashDataSourceConfigBasic(env.ID, string(targetNameOption)),
+			Check:  resource.TestCheckResourceAttrSet(dataSourceName, "script"),
+		}
+	}
+	// Make a step per valid targetNameOption
+	var steps []resource.TestStep
+	for _, opt := range targetnameoption.TargetNameOptionValues() {
+		steps = append(steps, makeTestStep(opt))
+	}
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: acctest.TestProtoV6ProviderFactories,
+		Steps:                    steps,
+	})
+}
+
+func TestADBashDataSource_InvalidEnvironmentID(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: acctest.TestProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				// Empty environment_id not permitted
+				Config:      testAccADBashDataSourceConfigBasic("", string(targetnameoption.BashHostName)),
```

Comment:
Question - I see we are testing the environmentId is invalid when empty. Do we also have validation for it being a valid GUID?

Reply:
The validation tests are checking that our defined local validators we add work correctly:

https://github.com/bastionzero/terraform-provider-bastionzero/blob/4c38dab3292e74e088f32121db2b4792fc1faa08/bastionzero/autodiscoveryscript/datasource_adbash.go#L30-L44

We could make a custom one to check for a valid GUID. However, if they pass an invalid GUID, it should throw a validation error on backend and we'll display that to the user.

```
Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Error reading autodiscovery script (bash)
│
│   with data.bastionzero_ad_bash.ad_script,
│   on main.tf line 107, in data "bastionzero_ad_bash" "ad_script":
│  107: data "bastionzero_ad_bash" "ad_script" {
│
│ GET https://cloud.bastionzero.com/api/v2/autodiscovery-scripts/bzero/bash?environmentId=foobar&targetNameOption=BashHostName: 400 Bad Request: environmentId: The value 'foobar' is not valid., A value for the 'environmentId' parameter or property was not provided.
╵
```

So now it's just a question of is this error friendly enough or do we want to write a custom one (and I guess also prevent the unnecessary API call by making it a local validator)?

Reply:
I see yeah i think it would be a little nicer if we did have a local validator and failed faster and with a better error message. If its not a crazy amount of work to add seems worth it since we use GUIDs throughout most of the API. Wouldnt it be as simple as adding a validator that used something like [this](https://stackoverflow.com/a/46315070). Its good that we are always bubbling up the error from the backend as a catch-all though.

Reply:
Oh yeah I can just use a regex for it. Nice!

Reply:
Or wait I'll just use that library. Apparently its faster and I think we already include it as a dep

Reply:
Added in d931139 to applicable data sources and resources that have an `id` that should be a UUIDv4 string.

Example error:

```
│ Error: Invalid Attribute Value Match
│
│   with data.bastionzero_bzero_target.bzt_test_timeout,
│   on main.tf line 34, in data "bastionzero_bzero_target" "bzt_test_timeout":
│   34:   id   = "foo"
│
│ Attribute id value must be a valid UUID v4 string, got: foo
╵
```